### ISSUE TITLE
Verify plausability of special phrase terms

### ISSUE THREAD

**Issue by lonvia** [osm-search/Nominatim#235]
The Special Phrase pages in the wiki get the occasional entry for tag combinations that don't really exist. Two things: a) make sure the key list is restricted to tags actually imported by osm2pgsql. b) Taginfo could be used to verify that a key/value combination actually exists.

**Comment by AntoJvlt** [comment]
It this issue still revelant @lonvia?

If I get your point well, during the import of the special phrases we should check if the key is present in the [generic_keys](https://github.com/openstreetmap/osm2pgsql/blob/f7993c95fc3575299418428e3ba07fb67162107a/style.lua#L28) array of  [osm2pgsql/style.lua](https://github.com/openstreetmap/osm2pgsql/blob/master/style.lua).

Then if it is present we could call the [TagInfo API](https://taginfo.openstreetmap.org/taginfo/apidoc) to check if the given value match with the tag.

For example if we get `aeroway=aerodrome`, we would call : 
[`https://taginfo.openstreetmap.org/api/4/tag/combinations?key=aeroway&value=aerodrome&format=json&rp=1&page=1`](https://taginfo.openstreetmap.org/api/4/tag/combinations?key=aeroway&value=aerodrome&format=json&rp=1&page=1)

And then check that `total` field is not equal to 0.

Off course for every keys and key/value pairs we would cache the result. This cache could be in memory during the import process or by creating a file to cache results inside, so that if we do another special phrases import we can use and update this cache.

I don't know if the second option is overkill and if there is a risk that a given key (or a key/value) may not be supported anymore or removed.

Please let me know what you think and how you imagined that, thanks.

**Comment by lonvia** [comment]
Yes, that's still relevant although for the moment I would keep to a). The taginfo idea would be more relevant if we wanted to start supplying precomputed special-terms lists instead of downloading the ones from the wiki, i.e. in that case, the actual computation of the special phrases would happen rather rarely.

By osm2pgsql config, Nominatim's style config is meant, e.g [this one](https://github.com/osm-search/Nominatim/blob/28b4fb12b684a5c76ca24002ba1ae19ce2e22c87/settings/import-extratags.style). [get_import_style_file()](https://nominatim.org/release-docs/latest/develop/Import/#configuring-the-import) gets you the configuration file in use. There is a [description of the file format](https://nominatim.org/release-docs/latest/develop/Import/#configuring-the-import) in the documentation. Essentially you would be interested in the key/value combinations with a 'main' property. Those combinations are the only ones that can be searched with the special terms later.

**Comment by lonvia** [comment]
Some updates on the issue as the code has moved a bit: The special phrase importer is now [part of the Python code](https://github.com/osm-search/Nominatim/blob/master/nominatim/tools/special_phrases/sp_importer.py). Implementing a) requires something like this:

1. After the [index on classes and types has been created](https://github.com/osm-search/Nominatim/blob/8191c747b964530f67b064b43954e13a5b1a0791/nominatim/tools/special_phrases/sp_importer.py#L180) query the database and get a list of combinations of `class` and `type` and their number of occurrences. Restrict the list to combinations with more than 100 occurrences.
2. Remove all class/type combinations from the list of phrases, which have not been found in the database.

**Comment by kitsiosvas** [comment]
@lonvia Can I have a bit more information about the problem? As far as I understand, you propose to create a new index on a database table? If so, which table in the data base are you referring to (also information about the table would be helpful, like column names). 

Also, are we talking here about new functionality in the tools/special_phrases/mp_importer.py file? Should this be a new function in it self?

**Comment by avinabneogy23** [comment]
Hi @lonvia , can you please assign this issue to me

**Comment by lonvia** [comment]
We don't assign issues here @avinabneogy23, just go ahead and give it a try. My comment from 2023 with guidance on the issue is still valid.

**Comment by Xclusive09** [comment]
Hi @lonvia , is this issue still valid ?
so i can get on with it

**Comment by avinabneogy23** [comment]
Hey @Xclusive09 , i did not have the time to look at it
Please feel free to get on with it
